File: src/modlunky2/ui/levels/shared/biomes.py
from modlunky2.mem.state import Theme

class BIOME:
    DWELLING = "cave"
    JUNGLE = "jungle"
    VOLCANA = "volcano"
    OLMEC = "olmec"
    TEMPLE = "temple"
    TIDE_POOL = "tidepool"
    ICE_CAVES = "ice"
    NEO_BABYLON = "babylon"
    SUNKEN_CITY = "sunken"
    BEEHIVE = "beehive"
    CITY_OF_GOLD = "gold"
    DUAT = "duat"
    EGGPLANT_WORLD = "eggplant"
    SURFACE = "surface"


class Biomes:
    @staticmethod
    def get_biome_for_level(
        lvl,
    ):  # cave by default, depicts what background and sprites will be loaded
        if (
            lvl.startswith("challenge_sun")
            or lvl.startswith("sunken")
            or lvl.startswith("hundun")
            or lvl.startswith("ending_hard")
            or lvl.endswith("_sunkencity.lvl")
        ):
            return BIOME.SUNKEN_CITY
        elif (
            lvl.startswith("abzu.lvl")
            or lvl.startswith("lake")
            or lvl.startswith("tide")
            or lvl.startswith("end")
            or lvl.endswith("_tidepool.lvl")
        ):
            return BIOME.TIDE_POOL
        elif (
            lvl.startswith("babylon")
            or lvl.startswith("hallofu")
            or lvl.endswith("_babylon.lvl")
            or lvl.startswith("palace")
            or lvl.startswith("tiamat")
        ):
            return BIOME.NEO_BABYLON
        elif lvl.startswith("basecamp"):
            return BIOME.DWELLING
        elif lvl.startswith("beehive"):
            return BIOME.BEEHIVE
        elif (
            lvl.startswith("blackmark")
            or lvl.startswith("jungle")
            or lvl.startswith("challenge_moon")
            or lvl.endswith("_jungle.lvl")
        ):
            return BIOME.JUNGLE
        elif (
            lvl.startswith("challenge_star")
            or lvl.startswith("temple")
            or lvl.endswith("_temple.lvl")
        ):
            return BIOME.TEMPLE
        elif lvl.startswith("city"):
            return BIOME.CITY_OF_GOLD
        elif lvl.startswith("duat"):
            return BIOME.DUAT
        elif lvl.startswith("egg"):
            return BIOME.EGGPLANT_WORLD
        elif lvl.startswith("ice") or lvl.endswith("_icecavesarea.lvl"):
            return BIOME.ICE_CAVES
        elif lvl.startswith("olmec"):
            return BIOME.OLMEC
        elif lvl.startswith("vlad"):
            return BIOME.VOLCANA
        elif lvl.startswith("volcano") or lvl.endswith("_volcano.lvl"):
            return BIOME.VOLCANA

        dm_themes = [
            BIOME.DWELLING,
            BIOME.JUNGLE,
            BIOME.VOLCANA,
            BIOME.TIDE_POOL,
            BIOME.TEMPLE,
            BIOME.ICE_CAVES,
            BIOME.NEO_BABYLON,
            BIOME.SUNKEN_CITY,
        ]
        for x, themeselect in enumerate(dm_themes):
            if lvl.startswith("dm" + str(x + 1)):
                return themeselect
        return BIOME.DWELLING

    # Used for selecting a theme to get the theme code that
    # corresponds to the display-friendly theme name.
    @staticmethod
    def biome_for_name(name):
        if name == "Dwelling":
            return BIOME.DWELLING
        elif name == "Jungle":
            return BIOME.JUNGLE
        elif name == "Volcana":
            return BIOME.VOLCANA
        elif name == "Olmec":
            return BIOME.OLMEC
        elif name == "Tide Pool":
            return BIOME.TIDE_POOL
        elif name == "Temple":
            return BIOME.TEMPLE
        elif name == "Ice Caves":
            return BIOME.ICE_CAVES
        elif name == "Neo Babylon":
            return BIOME.NEO_BABYLON
        elif name == "Sunken City":
            return BIOME.SUNKEN_CITY
        elif name == "City of Gold":
            return BIOME.CITY_OF_GOLD
        elif name == "Duat":
            return BIOME.DUAT
        elif name == "Eggplant World":
            return BIOME.EGGPLANT_WORLD
        elif name == "Surface":
            return BIOME.SURFACE
        return None

    # Gets a string that can be used to display the name of a biome.
    @staticmethod
    def name_of_biome(theme):
        if theme == BIOME.DWELLING:
            return "Dwelling"
        elif theme == BIOME.TIDE_POOL:
            return "Tide Pool"
        elif theme == BIOME.NEO_BABYLON:
            return "Neo Babylon"
        elif theme == BIOME.JUNGLE:
            return "Jungle"
        elif theme == BIOME.TEMPLE:
            return "Temple"
        elif theme == BIOME.SUNKEN_CITY:
            return "Sunken City"
        elif theme == BIOME.CITY_OF_GOLD:
            return "City of Gold"
        elif theme == BIOME.DUAT:
            return "Duat"
        elif theme == BIOME.EGGPLANT_WORLD:
            return "Eggplant World"
        elif theme == BIOME.ICE_CAVES:
            return "Ice Caves"
        elif theme == BIOME.OLMEC:
            return "Olmec"
        elif theme == BIOME.VOLCANA:
            return "Volcana"
        elif theme == BIOME.SURFACE:
            return "Surface"
        return "Unknown"

    @staticmethod
    def theme_for_biome(biome):
        if biome == BIOME.DWELLING:
            return Theme.DWELLING
        elif biome == BIOME.JUNGLE:
            return Theme.JUNGLE
        elif biome == BIOME.VOLCANA:
            return Theme.VOLCANA
        elif biome == BIOME.OLMEC:
            return Theme.OLMEC
        elif biome == BIOME.TIDE_POOL:
            return Theme.TIDE_POOL
        elif biome == BIOME.TEMPLE:
            return Theme.TEMPLE
        elif biome == BIOME.ICE_CAVES:
            return Theme.ICE_CAVES
        elif biome == BIOME.NEO_BABYLON:
            return Theme.NEO_BABYLON
        elif biome == BIOME.SUNKEN_CITY:
            return Theme.SUNKEN_CITY
        elif biome == BIOME.CITY_OF_GOLD:
            return Theme.CITY_OF_GOLD
        elif biome == BIOME.DUAT:
            return Theme.DUAT
        elif biome == BIOME.EGGPLANT_WORLD:
            return Theme.EGGPLANT_WORLD
        elif biome == BIOME.SURFACE:
            return Theme.BASE_CAMP
        return Theme.DWELLING

    @staticmethod
    def biome_for_theme(theme, subtheme):
        if theme == Theme.DWELLING:
            return BIOME.DWELLING
        elif theme == Theme.JUNGLE:
            return BIOME.JUNGLE
        elif theme == Theme.VOLCANA:
            return BIOME.VOLCANA
        elif theme == Theme.OLMEC:
            return BIOME.OLMEC
        elif theme == Theme.TIDE_POOL or theme == Theme.ABZU:
            return BIOME.TIDE_POOL
        elif theme == Theme.TEMPLE:
            return BIOME.TEMPLE
        elif theme == Theme.ICE_CAVES:
            return BIOME.ICE_CAVES
        elif theme == Theme.NEO_BABYLON or theme == Theme.TIAMAT:
            return BIOME.NEO_BABYLON
        elif theme == Theme.SUNKEN_CITY or theme == Theme.HUNDUN:
            return BIOME.SUNKEN_CITY
        elif theme == Theme.CITY_OF_GOLD:
            return BIOME.CITY_OF_GOLD
        elif theme == Theme.DUAT:
            return BIOME.DUAT
        elif theme == Theme.EGGPLANT_WORLD:
            return BIOME.EGGPLANT_WORLD
        elif theme == Theme.BASE_CAMP:
            return BIOME.SURFACE
        elif theme == Theme.ARENA or theme == Theme.COSMIC_OCEAN:
            if subtheme is not None:
                return Biomes.biome_for_theme(subtheme, None)
        return BIOME.DWELLING

File: app/src/components/levels/biomes.ts
// Filename / theme-id → biome resolver. Ports Python's
// `src/modlunky2/ui/levels/shared/biomes.py::Biomes` so both editors
// (Vanilla + Custom) pick the right floor tint and background PNG.
// Strings match Python's BIOME constants byte-for-byte so
// `getBiomeBackground(biome)` and `buildTileNameAtlas(names, biome)`
// on the Rust side (`FLOOR_BIOMES` in `level_editor.rs`) accept them
// without translation.

export type Biome =
  | "cave"
  | "jungle"
  | "volcano"
  | "olmec"
  | "temple"
  | "tidepool"
  | "ice"
  | "babylon"
  | "sunken"
  | "beehive"
  | "gold"
  | "duat"
  | "eggplant"
  | "surface";

/** Fallback used when no matcher / mapping fires. Matches Python's
 *  behavior of ending `Biomes.get_biome_for_level` with `return
 *  BIOME.DWELLING`. */
export const DEFAULT_BIOME: Biome = "cave";

/** Public-facing name for each biome. `cave` is the internal id for the
 *  Dwelling theme, `volcano` for Volcana, etc.; the UI should never surface
 *  the internal strings. */
export const BIOME_LABEL: Record<Biome, string> = {
  cave: "Dwelling",
  jungle: "Jungle",
  volcano: "Volcana",
  olmec: "Olmec",
  temple: "Temple",
  tidepool: "Tide Pool",
  ice: "Ice Caves",
  babylon: "Neo Babylon",
  sunken: "Sunken City",
  beehive: "Beehive",
  gold: "City of Gold",
  duat: "Duat",
  eggplant: "Eggplant World",
  surface: "Base Camp",
};

/** Representative theme id (from `LevelConfigPanel.THEMES`) for each biome.
 *  Inverse of `biomeForThemeId` for the non-ambiguous cases; used to turn a
 *  filename-derived biome back into a theme id (e.g. resolving the Cosmic
 *  Ocean subtheme of a `cosmicocean_*` file). Biomes with no dedicated theme
 *  (beehive) map to their closest relative and never surface as a subtheme. */
export const THEME_ID_FOR_BIOME: Record<Biome, number> = {
  cave: 1,
  jungle: 2,
  volcano: 3,
  olmec: 4,
  tidepool: 5,
  temple: 6,
  ice: 7,
  babylon: 8,
  sunken: 9,
  gold: 11,
  duat: 12,
  eggplant: 15,
  surface: 17,
  beehive: 2,
};

/** Deathmatch arena filenames use a `dm{N}` prefix where N is a
 *  1-based index into this ordered biome list. Matches Python's
 *  `dm_themes` array in `Biomes.get_biome_for_level`. */
const DM_THEMES: Biome[] = [
  "cave",
  "jungle",
  "volcano",
  "tidepool",
  "temple",
  "ice",
  "babylon",
  "sunken",
];

/** Filename → biome. Direct port of Python's
 *  `Biomes.get_biome_for_level(lvl)` in
 *  `src/modlunky2/ui/levels/shared/biomes.py`. Case-insensitive on the
 *  matches Python does with `startswith`/`endswith`.
 *
 *  Returns "cave" for anything the matcher doesn't recognize (basecamp
 *  and unknown filenames both land here on purpose). */
export function biomeForLevelFilename(fileName: string): Biome {
  const lvl = fileName.toLowerCase();

  if (
    lvl.startsWith("challenge_sun") ||
    lvl.startsWith("sunken") ||
    lvl.startsWith("hundun") ||
    lvl.startsWith("ending_hard") ||
    lvl.endsWith("_sunkencity.lvl")
  ) {
    return "sunken";
  }
  if (
    lvl.startsWith("abzu.lvl") ||
    lvl.startsWith("lake") ||
    lvl.startsWith("tide") ||
    lvl.startsWith("end") ||
    lvl.endsWith("_tidepool.lvl")
  ) {
    return "tidepool";
  }
  if (
    lvl.startsWith("babylon") ||
    lvl.startsWith("hallofu") ||
    lvl.endsWith("_babylon.lvl") ||
    lvl.startsWith("palace") ||
    lvl.startsWith("tiamat")
  ) {
    return "babylon";
  }
  if (lvl.startsWith("basecamp")) return "cave";
  if (lvl.startsWith("beehive")) return "beehive";
  if (
    lvl.startsWith("blackmark") ||
    lvl.startsWith("jungle") ||
    lvl.startsWith("challenge_moon") ||
    lvl.endsWith("_jungle.lvl")
  ) {
    return "jungle";
  }
  if (
    lvl.startsWith("challenge_star") ||
    lvl.startsWith("temple") ||
    lvl.endsWith("_temple.lvl")
  ) {
    return "temple";
  }
  if (lvl.startsWith("city")) return "gold";
  if (lvl.startsWith("duat")) return "duat";
  if (lvl.startsWith("egg")) return "eggplant";
  if (lvl.startsWith("ice") || lvl.endsWith("_icecavesarea.lvl")) return "ice";
  if (lvl.startsWith("olmec")) return "olmec";
  if (lvl.startsWith("vlad")) return "volcano";
  if (lvl.startsWith("volcano") || lvl.endsWith("_volcano.lvl")) return "volcano";

  for (let i = 0; i < DM_THEMES.length; i++) {
    if (lvl.startsWith(`dm${i + 1}`)) return DM_THEMES[i];
  }

  return DEFAULT_BIOME;
}

/** Theme id (from `LevelConfigPanel.THEMES`) → biome. Ports Python's
 *  `Biomes.biome_for_theme(theme, subtheme)`. Abzu falls back to Tide
 *  Pool, Tiamat to Neo Babylon, Hundun to Sunken City (they render the
 *  same biome art in the tkinter editor). Cosmic Ocean uses its
 *  subtheme when provided; when the subtheme is missing or Cosmic
 *  Ocean itself, returns "cave" (Python's ultimate fallback). */
export function biomeForThemeId(
  themeId: number | undefined,
  subthemeId?: number,
): Biome {
  switch (themeId) {
    case 1:
      return "cave"; // Dwelling
    case 2:
      return "jungle"; // Jungle
    case 3:
      return "volcano"; // Volcana
    case 4:
      return "olmec"; // Olmec
    case 5:
    case 13:
      return "tidepool"; // Tide Pool / Abzu
    case 6:
      return "temple"; // Temple
    case 7:
      return "ice"; // Ice Caves
    case 8:
    case 14:
      return "babylon"; // Neo Babylon / Tiamat
    case 9:
    case 16:
      return "sunken"; // Sunken City / Hundun
    case 10: {
      // Cosmic Ocean: recurse on the subtheme, otherwise cave.
      if (subthemeId != null && subthemeId !== 10) {
        return biomeForThemeId(subthemeId);
      }
      return "cave";
    }
    case 11:
      return "gold"; // City of Gold
    case 12:
      return "duat"; // Duat
    case 15:
      return "eggplant"; // Eggplant World
    case 17:
      return "surface"; // Base Camp
    default:
      return "cave";
  }
}
